text/0028-netowrk-collections-and-variables.md

```diff
@@ -0,0 +1,94 @@
+
+# Motivation
+[motivation]: #motivation
+
+The current programming model allows for unsafe, race-condition-prone operations.  Its design also in its current form is not readily adaptable to future snapshot plans.  Lastly, the goals of the library to be a performant netcode system are in conflict with some of its lack of limitations
+
+
+# Guide-Level Explanation
+[guide-level]: #guid-level
+
+## Permission Changes
+
+Today, the library has 3 levels of permissions
+
+### Old: Read & Write Permissions
+
+* `ServerOnly`
+* `OwnerOnly`
+* `Everyone`
+
+
+### New: Read Permissions
+* `Everyone`
+* `ServerOnly`
+
+Going forward, Network Variables will only be (directly) writable by the server.  Previously, a client could change state on the server by writing to its local network variable instance.  As we move towards the snapshot tick-based model this free-form mechanism becomes intractable to support at least in its current form.  With this change, when a client wants to cause a state change it will need to use a different means such as sending an RPC.  
+
+Hence, with this release there are only *Read* permissions.  Either a variable can be read by everyone, or the specific owner of the variable.  The server can always read all variables.
+
+The `Everyone` mode has also been removed.  Not only did this have the same issues as the client-writable mode, but also it led to race conditions when multiple clients and the server could write to a variable.
+
+## Change in network variable writing interface
+
+Currently the NetworkVariable class has a `Value` getter/setter.  When the `get` method is called, the internal, current value is returned.  When the `set` method is called the intention is that the value change is wrapped by code to dirty the variable and cause it to be updated.  This works great for scalar types, e.g.
+
+```
+NetworkVariable<float> someVar = new NetworkVariable<float>();
+someVar.Value = 1.23;
+
+```
+
+But not so great for struct access.
+
+```
+public class TestClass : INetworkSerializable
+{
+    public uint SomeFloat;
+    public bool SomeInt;
+
+    public void NetworkSerialize(NetworkSerializer serializer) { ... }
+}
+
+NetworkVariable<TestClass> TheClass = new NetworkVariable<TestClass>(new TestClass());
+TheClass.Value.SomeBool = false;
+TheClass.Value.SomeInt = k_TestUInt;
+```
+
+This code in fact causes nothing to happen on the server as it isn't in fact calling `set`.  Users must remember to call `TheClass.SetDirty(true)`.  
+
+Moreover, the above code doesn't work at all with struct member access.  With structs, the only way to change value is to assign a new copy.
+
+Instead, with this RFC network variables will no longer have a `set` on the `Value` property.  The `get` property will return the value by reference and mark the variable as dirty.  There will be mitigation around avoiding the fact that dirty-ing will happen more often to avoid unnecessary updates.
```

Comment:
Why mark something dirty if you are just reading it?  Is there an underlying logic behind this I am missing?